### ISSUE TITLE
Can we have a ResourceCollection() under Resource()?

### ISSUE THREAD

**Issue by vinaysharma77** [scott-mcdonald/JsonApiFramework#27]
.Net version 4.5.

I have an object called Process and under that I have collection of Activities

{
"processDocument": [
{
"jsonapi": {"version": "1.1"},
"data": {
"type": "process-model",
"id": "510003",
"attributes": {
"process-id": 510003,
"process-name": "Process123",
"enabled": true,
"activities": [
{
"processId": 510003,
"activityId": 510005,
"activityName": "Activity 1"
},
{
"processId": 510003,
"activityId": 510054,
"activityName": "Activity 2"
},
{
"processId": 510003,
"activityId": 510004,
"activityName": "Activity 3"
}
]
}
}
}
]
}

Here's what I'm trying to accomplish:

I want to be able to get the fully qualified links for each item in the activity collection. I'm able to output the above json without links by this code:

           var processdoc = documentContext
                .NewDocument(HttpContext.Current.Request.Url)
                    .SetJsonApiVersion(JsonApiVersion.Version11)
                    .Resource(process)
                    .ResourceEnd()
                .WriteDocument();
The "process" object supplied to the Resource() method above contains the Process properties and a collection (List) of activities. The JSON Api doesn't seem to support accessing collection the ResourceCollection() under Resource().. I was expecting to do something like this:

           var processdoc = documentContext
                .NewDocument(HttpContext.Current.Request.Url)
                    .SetJsonApiVersion(JsonApiVersion.Version11)
                    .Resource(process)                            
                        .ResourceCollection(process.Activities)
                            .Links()
                                .AddSelfLink()
                            .LinksEnd()
                        .ResourceCollectionEnd()
                    .ResourceEnd()
                .WriteDocument();
Or is there some other way to accomplish this using JSON Api that I'm not aware of?

I was also hoping to convert the property name to camel case (processId, processName) instead of using hyphens (process-id, process-name) in the JSON output. Is that doable in JSON Api?

Appreciate any help on these.

Vin

**Comment by scott-mcdonald** [comment]
@vinaysharma77 The quick answer is json:api resources can not be composed of other resources. Instead they have relationships to each other. See [Clarification on Resources Composed of Other Resources](http://discuss.jsonapi.org/t/clarification-on-resources-composed-of-other-resources/252) in the json:api forum for details on this topic. A corollary to this is do not include "foreign keys" in the resources, instead favor relationships.

So for you specific case instead of having process being composed of many activities, process has a "to-many" relationship to activities where the hypermedia API server can server-side "include" the related activities if desired or if the client requests with a "include" query parameter.

**Comment by scott-mcdonald** [comment]
Closing this issue as resources can not be composed of other resources.

**Comment by scott-mcdonald** [comment]
Here is an example how to create a document with the primary data being a Process resource and server-side including the related Activities:

```csharp
var process = GetProcess();
var processToManyActivities = GetProcessToManyActivities(process);
var processToManyActivitiesResourceLinkage = ToManyResourceLinkage.Create(process , "activities", processToManyActivities);

var document = documentContext
    .NewDocument(HttpContext.Current.Request.Url)
        .Links()
            .AddUpLink()
            .AddSelfLink()
        .LinksEnd()
        .Resource(process)
            .Relationships()
                .AddRelationship("activities", Keywords.Related, Keywords.Self)
            .RelationshipsEnd()
            .Links()
                .AddSelfLink()
            .LinksEnd()
        .ResourceEnd()
        .Included()
            .ToMany(processToManyActivitiesResourceLinkage)
                .Links()
                    .AddSelfLink()
                .LinksEnd()
            .ToManyEnd()
        .IncludedEnd()
    .WriteDocument();

or another version with a short-cut for adding the related resources

var document = documentContext
    .NewDocument(HttpContext.Current.Request.Url)
        .Links()
            .AddUpLink()
            .AddSelfLink()
        .LinksEnd()
        .Resource(process)
            .Relationships()
                .AddRelationship("activities", Keywords.Related, Keywords.Self)
            .RelationshipsEnd()
            .Links()
                .AddSelfLink()
            .LinksEnd()
        .ResourceEnd()
        .Included()
            .ToMany(process , "activities", processToManyActivities)
                .Links()
                    .AddSelfLink()
                .LinksEnd()
            .ToManyEnd()
        .IncludedEnd()
    .WriteDocument();

return this.Ok(document);

````
Something like this is what I believe you are looking for.

Good luck and if you like JsonApiFramework, please star this GitHub project.
/Scott/

**Comment by vinaysharma77** [comment]
@scott-mcdonald Thanks! The "Included()" did do what I wanted and I can see the activities collection in the JSON response now. But the URL that's getting built isn't correct. This is what I get:

This is my input url:

http://<server_name>/<virtual_dir_name>/api/workflow/process123

Response:

{
  "processDocument": [
    {
      "jsonapi": {
        "version": "1.1"
      },
      "data": {
        "type": "process-detail-models",
        "id": "process123",
        "attributes": {
          "process-id": 510003,
          "process-name": "Process 123",
          "enabled": true,
        },
		"relationships": {
        "activities": {
          "links": {
            "related": "http://<server_name>/<virtual_dir_name>/api/process123/workflow/process123/activities"
          },
          "data": [
            {
              "type": "activity-models",
              "id": "activeleads"
            },
            {
              "type": "activity-models",
              "id": "csvduplicatelead"
            }            
          ]
         }
       },
	   "links": {
        "self": "http://<server_name>/<virtual_dir_name>/api/process123/workflow/process123"
       }
      },
	  "included": [
      {
        "type": "activity-models",
        "id": "activity1",
        "attributes": {
          "activity-id": 510005,
          "activity-name": "Activity 1"
        },
        "links": {
          "self": "http://<server_name>/<virtual_dir_name>/api/process123/activity-models/activity1"
        }
      },
      {
        "type": "activity-models",
        "id": "activity2",
        "attributes": {
          "activity-id": 510054,
          "activity-name": "CSV Duplicate Lead"
        },
        "links": {
          "self": "http://<server_name>/<virtual_dir_name>/api/process123/activity-models/activity2"
        }
      }
    }
  ]
}

The correct links for the process (Process 123) and its respective activities (Activity 1 and Activity 2) is:

Process 123 --> http://<server_name>/<virtual_dir_name>/api/workflow/process123
Activity 1 --> http://<server_name>/<virtual_dir_name>/api/workflow/process123/activity1
Activity 1 --> http://<server_name>/<virtual_dir_name>/api/workflow/process123/activity2

As you can see I get process name two times in the self url generated for the process and the segment "workflow" is missing in the activity url. Instead it has "activity-models" in the URL (which I believe is coming from the name of my Activity POCO).

I tried adding "Hypermedia().SetApiCollectionPathSegment("workflow");" in my activity configuration to replace the POCO name with "workflow" (as I did for Process) but get a run time exception:

**An item with the same key has already been added.**

Stack Trace:
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at JsonApiFramework.Server.Hypermedia.Internal.UriExtensions.InitialIteration(IServiceModel serviceModel, IEnumerator`1 urlPathSegmentsEnumerator, ICollection`1 documentSelfPath) in C:\Research\JsonApiFramework-master\Source\JsonApiFramework.Server\Hypermedia\Internal\UriExtensions.cs:line 83
   at JsonApiFramework.Server.Hypermedia.Internal.UriExtensions.ParseDocumentSelfPath(Uri url, IServiceModel serviceModel, IUrlBuilderConfiguration urlBuilderConfiguration) in... 

Any insight on how to get the URIs correct would be really helpful as that's one of our primary objective to explore JSON Api.

Thanks again.

Vinay

**Comment by scott-mcdonald** [comment]
I would need to see the WebAPI project, can I clone it in GitHub? I'll take a look if I can see the source code, or at least source code that reproduces this issue.

**Comment by vinaysharma77** [comment]
Scott,

There were some realignment of priorities hence  couldn't work on getting a test app ready for you. Apologies for the delay.

I now have a git repository of my project that reproduces the issue. Here's the link: https://github.com/vinaysharma77/jsonapitest

you'll find 3 get routes in the WorkflowController:

api/workflow
api/workflow/{processName}
api/workflow/{processName}/{activityName}

The URL creation for "api/workflow" call works fine but when I get a collection (included) of activities in "api/workflow/{processName}" the self link URL created for the process and the link for the activities in the collection returned are not accurate. This is the response I get for "" call as you can see the segments in the URL returned are duplicated and not in the correct order as supplied in the request URL:

URL generated for the process is:

"http://<server_name>/jsonapitest/api/process1/workflow/process1"

It should be:

"http://<server_name>/jsonapitest/api/workflow/process1"

URL generated for the activities is:

"http://<server_name>/jsonapitest/api/process1/activity-models/activity1"

It should be:

"http://<server_name>/jsonapitest/api/workflow/process1/activity1"


{
  "processDocument": {
    "jsonapi": {
      "version": "1.1"
    },
    "data": {
      "type": "process-detail-models",
      "id": "process1",
      "attributes": {
        "process-id": 1,
        "process-name": "Process 1",
        "process-description": "This is process 1",
        "hide-on-ui": false,
        "process-route-name": "process1"
      },
      "relationships": {
        "activities": {
          "links": {
            "related": "http://<server_name>/jsonapitest/api/process1/workflow/process1/activities"
          },
          "data": [
            {
              "type": "activity-models",
              "id": "activity1"
            },
            {
              "type": "activity-models",
              "id": "activity2"
            },
            {
              "type": "activity-models",
              "id": "activity3"
            }
          ]
        }
      },
      "links": {
        "self": "http://<server_name>/jsonapitest/api/process1/workflow/process1"
      }
    },
    "included": [
      {
        "type": "activity-models",
        "id": "activity1",
        "attributes": {
          "activity-id": 1,
          "activity-name": "Activity 1",
          "activity-desc": "This is Activity 1 of Process 1",
          "activity-route-name": "activity1"
        },
        "links": {
          "self": "http://<server_name>/jsonapitest/api/process1/activity-models/activity1"
        }
      },
      {
        "type": "activity-models",
        "id": "activity2",
        "attributes": {
          "activity-id": 2,
          "activity-name": "Activity 2",
          "activity-desc": "This is Activity 2 of Process 1",
          "activity-route-name": "activity2"
        },
        "links": {
          "self": "http://<server_name>/jsonapitest/api/process1/activity-models/activity2"
        }
      },
      {
        "type": "activity-models",
        "id": "activity3",
        "attributes": {
          "activity-id": 3,
          "activity-name": "Activity 3",
          "activity-desc": "This is Activity 3 of Process 1",
          "activity-route-name": "activity3"
        },
        "links": {
          "self": "http://<server_name>/jsonapitest/api/process1/activity-models/activity3"
        }
      }
    ]
  }
}

**Comment by vinaysharma77** [comment]
Also, I downloaded the Blogs and Articles sample from this location https://github.com/scott-mcdonald/JsonApiFramework.Samples and I noticed that the URL returned for the articles for the blog in response to this request - "http://<server_name>/jsonapi/blogs" are also not accurate:

The Articles' URL is missing the "virtual directory" name (jsonapi)

I get this in the response:

"http://localhost:80/blogs/1/articles"

It should be:

"http://localhost:80/jsonapi/blogs/1/articles"



{
  "jsonapi": {
    "version": "1.0"
  },
  "links": {
    "up": "http://localhost:80/jsonapi",
    "self": "http://localhost:80/jsonapi/blogs"
  },
  "data": [
    {
      "type": "blogs",
      "id": "1",
      "attributes": {
        "name": "Et"
      },
      "relationships": {
        "articles": {
          "links": {
            "related": "http://localhost:80/blogs/1/articles"
          }
        }
      },
      "links": {
        "self": "http://localhost:80/blogs/1"
      }
    },
    {
      "type": "blogs",
      "id": "2",
      "attributes": {
        "name": "Stet Sea Nonumy Nonumy Qui Amet Congue"
      },
      "relationships": {
        "articles": {
          "links": {
            "related": "http://localhost:80/blogs/2/articles"
          }
        }
      },......

Please let me know if I'm missing something from the JSON Api usability standpoint.

Thanks.

**Comment by scott-mcdonald** [comment]
I have a feature branch for my version of the test project your created but I do not have permission to push to origin this feature branch. You do some things in an unorthodox manner, especially the final document having an extra level such as having "Processes" and "ProcessDocument" as the only property of the root object followed by the build json:api document as the property value - not standard json:api but I am not prevy to exactly what you are doing.

In my version, I had trouble with your WebApi project and the razor views so I created an much simpler, minimal Web API only project with no MVC to simplify the overall project.

I improved the building of the "UrlBuilderConfiguration" so the framework URL building was correct with your use of a virtual directory path (jsonapitest) and a convention based path (api).

The path building I see is what I would expect the paths to be. So take a look and that should give you a start on what you are trying to do.

Yes, you did catch a bug in the samples if you use virtual directory paths when hosting the Web API services so good catch.

Regards,
Scott
